File: packages/typescript/src/semantics.ts
import { fail } from './errors.js';
import type { JsonObject } from './json.js';
import { arrayAt, objectAt } from './paths.js';

function requireCondition(condition: boolean, path: string, message: string): void {
  if (!condition) fail('E_SEMANTIC', message, path);
}

function increasing(values: number[]): boolean {
  return values.every((value, index) => index === 0 || values[index - 1]! < value);
}

/** Cross-field rules that can be checked without a raster dataset or renderer. */
export function validateSemantics(style: JsonObject): void {
  const channelSpec = objectAt(style, 'renderer');
  const channels =
    'bidx' in channelSpec
      ? arrayAt(channelSpec, 'bidx').length
      : 'expression' in channelSpec
        ? (channelSpec.expression as string).split(';').length
        : 1;
  const renderer = channelSpec;
  const type = renderer.type;
  requireCondition(
    channels === (type === 'rgb' ? 3 : 1),
    'kind',
    'Renderer channel cardinality mismatch',
  );
  const stretch = objectAt(style, 'stretch');
  for (const key of ['rescale', 'curves']) {
    if (key in stretch) {
      const count = arrayAt(stretch, key).length;
      requireCondition(
        count === 1 || count === channels,
        key,
        'Expected one value or one per channel',
      );
    }
  }
  for (const pair of arrayAt(stretch, 'rescale') as number[][]) {
    requireCondition(pair[0]! < pair[1]!, 'rescale', 'Range must increase');
  }
  const percentiles = stretch.percentiles as number[] | undefined;
  if (percentiles)
    requireCondition(percentiles[0]! < percentiles[1]!, 'percentiles', 'Percentiles must increase');
  for (const curve of arrayAt(stretch, 'curves') as number[][][]) {
    requireCondition(
      curve.every((point) => point[1]! >= 0 && point[1]! <= 1),
      'curves',
      'Curve output must be in [0,1]',
    );
    requireCondition(
      increasing(curve.map((point) => point[0]!)) &&
        curve.every((point, index) => index === 0 || curve[index - 1]![1]! <= point[1]!),
      'curves',
      'Curve must be monotonic',
    );
  }
  const colorMap = objectAt(renderer, 'color_mapping');
  const bypass =
    ['categorized', 'single_color', 'hillshade'].includes(type as string) ||
    colorMap.domain === 'data';
  const nativeData =
    'colormap' in renderer &&
    (stretch.method ?? 'none') === 'none' &&
    !('color_formula' in objectAt(style, 'effects'));
  if (nativeData)
    requireCondition(
      !('range_policy' in stretch),
      'stretch.range_policy',
      'Native data colormaps use their own boundaries',
    );
  if (bypass) {
    requireCondition(
      (stretch.method ?? 'none') === 'none' && !('color_formula' in objectAt(style, 'effects')),
      'method',
      'Data-domain renderer must bypass stretch',
    );
    requireCondition(
      !('range_policy' in stretch),
      'stretch.range_policy',
      'Range policy requires a display-domain stretch',
    );
  }
  if (colorMap.mode === 'continuous' && 'stops' in colorMap) {
    const values = (colorMap.stops as JsonObject[]).map((stop) => stop.value as number);
    requireCondition(increasing(values), 'color_mapping.stops', 'Stops must increase');
    if (colorMap.domain === 'normalized')
      requireCondition(
        values.every((value) => value >= 0 && value <= 1),
        'color_mapping.stops',
        'Normalized stops must be in [0,1]',
      );
  }
  if (colorMap.mode === 'discrete') {
    const breaks = colorMap.breaks as number[];
    requireCondition(
      arrayAt(colorMap, 'colors').length + 1 === breaks.length,
      'color_mapping',
      'Break/color cardinality mismatch',
    );
    requireCondition(increasing(breaks), 'color_mapping.breaks', 'Breaks must increase');
  }
  if (colorMap.mode === 'exact') {
    const values = (colorMap.entries as JsonObject[]).map((entry) => entry.value);
    requireCondition(
      new Set(values).size === values.length,
      'color_mapping.entries',
      'Exact keys must be unique',
    );
  }
  const mosaic = objectAt(style, 'mosaic');
  if (type === 'categorized') {
    requireCondition(
      [
        objectAt(style, 'resampling').read ?? 'nearest',
        objectAt(style, 'resampling').reproject ?? 'nearest',
      ].every((value) => value === 'nearest' || value === 'mode'),
      'resampling',
      'Categorical rendering requires nearest or mode',
    );
    requireCondition(
      !['mean', 'median'].includes(mosaic.pixel_selection as string),
      'pixel_selection',
      'Categorical rendering forbids arithmetic mosaic',
    );
  }
  const calibration = objectAt(style, 'calibration');
  if (colorMap.mode === 'source') {
    requireCondition(
      (calibration.mode ?? 'none') === 'none',
      'calibration.mode',
      'Source palettes require unmodified category values',
    );
    requireCondition(
      Object.values(objectAt(style, 'extensions')).every(
        (extension) => (extension as JsonObject).stage === 'after_color',
      ),
      'extensions',
      'Source palettes allow only after_color extensions',
    );
  }
  if (calibration.mode === 'linear') {
    const bands = (calibration.coefficients as JsonObject[]).map((coefficient) => coefficient.band);
    requireCondition(
      new Set(bands).size === bands.length,
      'calibration',
      'Duplicate calibration band',
    );
  }
  if ('rank_channel' in mosaic) {
    requireCondition(
      ['highest', 'lowest'].includes(mosaic.pixel_selection as string),
      'rank_channel',
      'Rank applies only to highest/lowest',
    );
    if (mosaic.stage === 'after_channels')
      requireCondition(
        (mosaic.rank_channel as number) <= channels,
        'rank_channel',
        'Rank channel is out of bounds',
      );
    else if ('bidx' in renderer || 'index' in renderer) {
      const inputs =
        'bidx' in renderer
          ? arrayAt(renderer, 'bidx')
          : Object.values(objectAt(objectAt(renderer, 'index'), 'bindings'));
      requireCondition(
        (mosaic.rank_channel as number) <= new Set(inputs).size,
        'rank_channel',
        'Rank channel exceeds the distinct input band count',
      );
    }
  }
  const image = objectAt(style, 'image');
  const format = image.format ?? 'png';
  if ('background' in image) {
    const color = image.background as string;
    requireCondition(
      color.length === 7 || color.slice(-2).toLowerCase() === 'ff',
      'background',
      'Background must be opaque',
    );
  }
  if (format === 'png')
    requireCondition(!('quality' in image), 'quality', 'PNG does not accept lossy quality');
  if ('lossless' in image)
    requireCondition(format === 'webp', 'lossless', 'Lossless is a WebP option');
  if (image.lossless === true)
    requireCondition(
      !('quality' in image),
      'quality',
      'Lossless output does not accept lossy quality',
    );
}

File: packages/go/semantics.go
package rasterstyle

import "strings"

func increasing(values []any) bool {
	for index := 1; index < len(values); index++ {
		if values[index-1].(float64) >= values[index].(float64) {
			return false
		}
	}
	return true
}

// validateSemantics runs only after structural schema validation succeeds.
func validateSemantics(style map[string]any) error {
	var firstError error
	require := func(condition bool, path, message string) {
		if !condition && firstError == nil {
			firstError = failure("E_SEMANTIC", message, path)
		}
	}
	channelSpec := objectAt(style, "renderer")
	channels := 1
	if bands, ok := channelSpec["bidx"]; ok {
		channels = len(bands.([]any))
	} else if expression, ok := channelSpec["expression"]; ok {
		channels = len(strings.Split(expression.(string), ";"))
	}
	renderer := channelSpec
	expected := 1
	if renderer["type"] == "rgb" {
		expected = 3
	}
	require(channels == expected, "kind", "Renderer channel cardinality mismatch")
	stretch := objectAt(style, "stretch")
	for _, key := range []string{"rescale", "curves"} {
		if values, exists := stretch[key]; exists {
			count := len(values.([]any))
			require(count == 1 || count == channels, key, "Expected one value or one per channel")
		}
	}
	for _, value := range arrayAt(stretch, "rescale") {
		pair := value.([]any)
		require(pair[0].(float64) < pair[1].(float64), "rescale", "Range must increase")
	}
	if pair, exists := stretch["percentiles"].([]any); exists {
		require(pair[0].(float64) < pair[1].(float64), "percentiles", "Percentiles must increase")
	}
	for _, value := range arrayAt(stretch, "curves") {
		curve := value.([]any)
		for index, value := range curve {
			point := value.([]any)
			require(point[1].(float64) >= 0 && point[1].(float64) <= 1, "curves", "Curve output must be in [0,1]")
			if index > 0 {
				previous := curve[index-1].([]any)
				require(previous[0].(float64) < point[0].(float64) && previous[1].(float64) <= point[1].(float64), "curves", "Curve must be monotonic")
			}
		}
	}
	colorMap := objectAt(renderer, "color_mapping")
	bypass := renderer["type"] == "categorized" || renderer["type"] == "single_color" || renderer["type"] == "hillshade" || colorMap["domain"] == "data"
	if bypass {
		method, present := stretch["method"]
		require(!present || method == "none", "method", "Data-domain renderer must bypass stretch")
		_, formula := objectAt(style, "effects")["color_formula"]
		require(!formula, "color_formula", "Data-domain renderer must bypass color formula")
	}
	switch colorMap["mode"] {
	case "continuous":
		values := make([]any, 0)
		for _, item := range arrayAt(colorMap, "stops") {
			value := item.(map[string]any)["value"].(float64)
			values = append(values, value)
			if colorMap["domain"] == "normalized" {
				require(value >= 0 && value <= 1, "color_mapping.stops", "Normalized stops must be in [0,1]")
			}
		}
		require(increasing(values), "color_mapping.stops", "Stops must increase")
	case "discrete":
		breaks := arrayAt(colorMap, "breaks")
		require(len(arrayAt(colorMap, "colors"))+1 == len(breaks), "color_mapping", "Break/color cardinality mismatch")
		require(increasing(breaks), "color_mapping.breaks", "Breaks must increase")
	case "exact":
		seen := make(map[float64]bool)
		for _, item := range arrayAt(colorMap, "entries") {
			value := item.(map[string]any)["value"].(float64)
			require(!seen[value], "color_mapping.entries", "Exact keys must be unique")
			seen[value] = true
		}
	}
	mosaic := objectAt(style, "mosaic")
	if renderer["type"] == "categorized" {
		for _, key := range []string{"read", "reproject"} {
			value := objectAt(style, "resampling")[key]
			require(value == nil || value == "nearest" || value == "mode", "resampling", "Categorical rendering requires nearest or mode")
		}
		require(mosaic["pixel_selection"] != "mean" && mosaic["pixel_selection"] != "median", "pixel_selection", "Categorical rendering forbids arithmetic mosaic")
	}
	calibration := objectAt(style, "calibration")
	seenBands := make(map[float64]bool)
	for _, item := range arrayAt(calibration, "coefficients") {
		band := item.(map[string]any)["band"].(float64)
		require(!seenBands[band], "calibration", "Duplicate calibration band")
		seenBands[band] = true
	}
	if rank, exists := mosaic["rank_channel"]; exists {
		require(mosaic["pixel_selection"] == "highest" || mosaic["pixel_selection"] == "lowest", "rank_channel", "Rank applies only to highest/lowest")
		if mosaic["stage"] == "after_channels" {
			require(rank.(float64) <= float64(channels), "rank_channel", "Rank channel is out of bounds")
		}
	}
	image := objectAt(style, "image")
	format := image["format"]
	if format == nil {
		format = "png"
	}
	if color, exists := image["background"].(string); exists {
		require(len(color) == 7 || strings.HasSuffix(strings.ToLower(color), "ff"), "background", "Background must be opaque")
	}
	_, quality := image["quality"]
	if format == "png" {
		require(!quality, "quality", "PNG does not accept lossy quality")
	}
	if _, lossless := image["lossless"]; lossless {
		require(format == "webp", "lossless", "Lossless is a WebP option")
	}
	if image["lossless"] == true {
		require(!quality, "quality", "Lossless output does not accept lossy quality")
	}
	return firstError
}
